cdap-common/src/main/java/io/cdap/cdap/common/leveldb/LevelDBClassRewriter.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.common.leveldb;
+
+import io.cdap.cdap.common.lang.ClassRewriter;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.commons.AdviceAdapter;
+import org.objectweb.asm.commons.Method;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.Executor;
+import javax.annotation.Nullable;
+
+/**
+ * Current java LevelDB implementation has a memory accumulation problem during compaction.
+ * Each {@link org.iq80.leveldb.impl.FileMetaData} object holds minimum and maximum keys for the file.
+ * And those keys may reside in a slice pointing to a byte array with the whole file. In this case
+ * instead of holding 20 bytes reference, we hold a multi-kb or even mb reference per file, leading to OOM
+ * on big compaction. The patch checks all {@link org.iq80.leveldb.impl.InternalKey} objects while they
+ * are created and if it received such heavy metadata reference, makes a copy of the key (usually few dozen bytes)
+ * and uses this lightweight metadata reference instead.
+ */
+public class LevelDBClassRewriter implements ClassRewriter {
+  /**
+   * Full filename of the {@link org.iq80.leveldb.impl.FileMetaData} class to be patched
+   */
+  private static final String FILE_META_DATA_CLASS_NAME = "org.iq80.leveldb.impl.FileMetaData";
+  /**
+   * Resource name of {@link FileMetaDataUtil} class that has normalization static methos to trim key storage down
+   */
+  private static final String FILE_META_DATA_UTIL_CLASS_RESOURCE = "io/cdap/cdap/common/leveldb/FileMetaDataUtil";
+  /**
+   * Type of {@link FileMetaDataUtil} class that has normalization static methos to trim key storage down
+   */
+  private static final Type FILE_META_DATA_UTIL_TYPE = Type.getObjectType(FILE_META_DATA_UTIL_CLASS_RESOURCE);
```

Comment:
You probably can just do a `Type.getType(FileMetaDataUtil.class);` since the util class is in the CDAP code base?

Reply:
That one refers leveldb classes and I wanted to ensure that this one can be loaded and #needRewrite called even without LevelDB in classpath. Let me test is if would still work

Reply:
It seems to work fine even without LevelDB in the classpath. Will update now

Reply:
Done